File: example02threading/src/main/java/foo/bar/example/forethreading/ui/CounterView.java
package foo.bar.example.forethreading.ui;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.ScrollView;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import co.early.fore.core.observer.Observer;
import foo.bar.example.forethreading.OG;
import foo.bar.example.forethreading.R;
import foo.bar.example.forethreading.feature.counter.CounterWithLambdas;
import foo.bar.example.forethreading.feature.counter.CounterWithProgress;

/**
 *
 */
public class CounterView extends ScrollView {

    //models that we need to sync with
    private CounterWithProgress counterWithProgress;
    private CounterWithLambdas counterWithLambdas;


    //UI elements that we care about
    @BindView(R.id.counterwprog_increase_btn)
    public Button increaseBy20Prog;

    @BindView(R.id.counterwprog_busy_progress)
    public ProgressBar busyProg;

    @BindView(R.id.counterwprog_progress_txt)
    public TextView progressNumberProg;

    @BindView(R.id.counterwprog_current_txt)
    public TextView currentNumberProg;

    @BindView(R.id.counterwlambda_increase_btn)
    public Button increaseBy20Basic;

    @BindView(R.id.counterwlambda_busy_progress)
    public ProgressBar busyBasic;

    @BindView(R.id.counterwlambda_current_txt)
    public TextView currentNumberBasic;



    //single observer reference
    Observer observer = this::syncView;



    public CounterView(Context context) {
        super(context);
    }

    public CounterView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CounterView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }



    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        ButterKnife.bind(this, this);

        getModelReferences();

        setupButtonClickListeners();
    }


    private void getModelReferences(){
        counterWithProgress = OG.get(CounterWithProgress.class);
        counterWithLambdas = OG.get(CounterWithLambdas.class);
    }

    private void setupButtonClickListeners() {
        increaseBy20Prog.setOnClickListener(v -> counterWithProgress.increaseBy20());
        increaseBy20Basic.setOnClickListener(v -> counterWithLambdas.increaseBy20());
    }


    //data binding stuff below

    public void syncView(){
        increaseBy20Prog.setEnabled(!counterWithProgress.isBusy());
        busyProg.setVisibility(counterWithProgress.isBusy() ? VISIBLE : INVISIBLE);
        progressNumberProg.setText("" + counterWithProgress.getProgress());
        currentNumberProg.setText("" + counterWithProgress.getCount());

        increaseBy20Basic.setEnabled(!counterWithLambdas.isBusy());
        busyBasic.setVisibility(counterWithLambdas.isBusy() ? VISIBLE : INVISIBLE);
        currentNumberBasic.setText("" + counterWithLambdas.getCount());
    }


    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        counterWithLambdas.addObserver(observer);
        counterWithProgress.addObserver(observer);
        syncView(); //<-- don't forget this
    }


    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        counterWithLambdas.removeObserver(observer);
        counterWithProgress.removeObserver(observer);
    }
}

File: example-kt-02coroutine/src/main/java/foo/bar/example/forecoroutine/ui/CounterView.kt
package foo.bar.example.forecoroutine.ui

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ScrollView
import foo.bar.example.forecoroutine.OG
import foo.bar.example.forecoroutine.feature.counter.Counter
import foo.bar.example.forecoroutine.feature.counter.CounterWithProgress
import kotlinx.android.synthetic.main.fragment_counter.view.counter_busy_progress
import kotlinx.android.synthetic.main.fragment_counter.view.counter_current_txt
import kotlinx.android.synthetic.main.fragment_counter.view.counter_increase_btn
import kotlinx.android.synthetic.main.fragment_counter.view.counterwprog_busy_progress
import kotlinx.android.synthetic.main.fragment_counter.view.counterwprog_current_txt
import kotlinx.android.synthetic.main.fragment_counter.view.counterwprog_increase_btn
import kotlinx.android.synthetic.main.fragment_counter.view.counterwprog_progress_txt

/**
 * Copyright © 2019 early.co. All rights reserved.
 */
class CounterView @JvmOverloads constructor(
        context: Context?,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : ScrollView(context, attrs, defStyleAttr) {


    //models that we need to sync with
    private lateinit var counterWithProgress: CounterWithProgress
    private lateinit var counter: Counter


    //single observer reference
    private var observer = this::syncView


    override fun onFinishInflate() {
        super.onFinishInflate()

        getModelReferences()

        setupButtonClickListeners()
    }

    private fun getModelReferences() {
        counter = OG[Counter::class.java]
        counterWithProgress = OG[CounterWithProgress::class.java]
    }

    private fun setupButtonClickListeners() {
        counter_increase_btn.setOnClickListener { counter.increaseBy20() }
        counterwprog_increase_btn.setOnClickListener { counterWithProgress.increaseBy20() }
    }


    //data binding stuff below

    fun syncView() {
        counterwprog_increase_btn.isEnabled = !counterWithProgress.isBusy
        counterwprog_busy_progress.visibility = if (counterWithProgress.isBusy) View.VISIBLE else View.INVISIBLE
        counterwprog_progress_txt.text = "${counterWithProgress.progress}"
        counterwprog_current_txt.text = "${counterWithProgress.count}"

        counter_increase_btn.isEnabled = !counter.isBusy
        counter_busy_progress.visibility = if (counter.isBusy) View.VISIBLE else View.INVISIBLE
        counter_current_txt.text = "${counter.count}"
    }


    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        counter.addObserver(observer)
        counterWithProgress.addObserver(observer)
        syncView() //<-- don't forget this
    }


    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        counter.removeObserver(observer)
        counterWithProgress.removeObserver(observer)
    }
}
